### ISSUE TITLE
support jinja-style macros

### ISSUE THREAD

**Issue by abelnation** [yogthos/Selmer#127]
docs:
http://jinja.pocoo.org/docs/dev/templates/#macros

**Comment by yogthos** [comment]
I'm open to a pr for this, but I likely wouldn't have a chance to work on adding macros in the near future.

**Comment by brettatoms** [comment]
Jinja-style macros are really helpful for reducing markup that gets repeated over and over.  This is how I use them:
```html
<!-- The icons.html file contains macros that return the markup for SVG icons -->
{% import "icons.html" as icon %}

{% macro sidebar_item(title, href, icon) -%}
  <li>
    <a href="{{ href }}"
       class="text-gray-600 hover:bg-gray-50 hover:text-gray-900 group flex items-center px-2 py-2 text-base font-medium rounded-md"
    >
      <div>{{ icon }}</div>
      {{ title }}
    </a>
  </li>
{%- endmacro %}

<ul>
  {{ sidebar_item("Home", "/home", icon.home() }}
  {{ sidebar_item("Users", "/users", icon.users() }}
  {{ sidebar_item("Organizations", "/organizations", icon.someicon() }}
</ul>

```

IMO this would be a killer feature for Selmer.

**Comment by boxxxie** [comment]
`icon` in this case is a macro

i'm already doing something pretty similar to this on the app i'm working on, which does a lot of meta-templating (templates that return templates)

`{%-` i have seen this syntax before, but i couldn't find a definition for it in the jinja docs

how do `{% import` and `{% extends` interact ?

what is the difference between `{% import` and `{% include` ? other than aliasing

**Comment by brettatoms** [comment]
> `icon` in this case is a macro

Yes `icon` is a macro.  So in this example we import a macro from another file and also define and use a macro (`sidebar_item`) that calls the `icon` macro.

> `{%-` i have seen this syntax before, but i couldn't find a definition for it in the jinja docs

The `{%-` tag strips whitespace: https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control

> how do `{% import` and `{% extends` interact ?

They're mutually exclusive.  You can only extend one other template but you can import multiple other templates.

> what is the difference between `{% import` and `{% include` ? other than aliasing

The different has to do with passing the current context.  See https://jinja.palletsprojects.com/en/3.1.x/templates/#import-context-behavior
